File: Sources/Ledgelings/ChatClient.swift
import Foundation

/// One language model behind an OpenAI-style chat endpoint. LM Studio on this
/// Mac and OpenRouter on the internet both speak that dialect, so one client
/// serves both; only the address, the key and the model differ.
///
/// Any feature that wants words from a model builds one of these (normally
/// through `AppSettings.chatClient()`) and calls `reply`. Banter is the first
/// such feature; it is not special here.
///
/// One trap this guards against: ask LM Studio for a model that is not
/// installed and it silently answers with whatever model happens to be loaded.
/// So `checkModel` compares the id against the server's list before any line
/// is asked for. OpenRouter refuses unknown ids by itself, but the same check
/// gives the settings window the same clear answer.
struct ChatClient: Sendable {
    enum Provider: String, CaseIterable, Codable, Sendable {
        case lmStudio, openRouter

        var title: String {
            switch self {
            case .lmStudio: "LM Studio"
            case .openRouter: "OpenRouter"
            }
        }
    }

    enum Failure: Error, CustomStringConvertible {
        case serverDown(String)
        case modelMissing(String, available: [String])
        case refused(String)
        case badReply(String)

        var description: String {
            switch self {
            case .serverDown(let why): "the server is not answering: \(why)"
            case .modelMissing(let model, let available):
                "model \(model) is not available" + (available.isEmpty ? "" : " (have: \(available.prefix(8).joined(separator: ", ")))")
            case .refused(let message): "the server refused: \(message)"
            case .badReply(let what): "unexpected reply: \(what)"
            }
        }
    }

    static let openRouterURL = URL(string: "https://openrouter.ai/api/v1")!

    var provider: Provider
    /// The `/v1` root: `http://localhost:1234/v1` or `https://openrouter.ai/api/v1`.
    var baseURL: URL
    var apiKey: String?
    var model: String
    var timeout: TimeInterval = 60

    static func lmStudio(server: URL, model: String) -> ChatClient {
        ChatClient(provider: .lmStudio, baseURL: server.appendingPathComponent("v1"), apiKey: nil, model: model)
    }

    static func openRouter(key: String, model: String) -> ChatClient {
        ChatClient(provider: .openRouter, baseURL: openRouterURL, apiKey: key, model: model)
    }

    /// Enough to read OpenRouter's public model list; it needs no key.
    static let openRouterPublic = ChatClient(provider: .openRouter, baseURL: openRouterURL, apiKey: nil, model: "")

    var modelsURL: URL { baseURL.appendingPathComponent("models") }

    // MARK: Requests and replies, no network

    func request(system: String, user: String, maxTokens: Int = 80, temperature: Double = 0.9) throws -> URLRequest {
        struct Body: Encodable {
            struct Message: Encodable { let role: String; let content: String }
            let model: String; let messages: [Message]; let temperature: Double; let max_tokens: Int
        }
        var request = authorised(URLRequest(url: baseURL.appendingPathComponent("chat/completions"), timeoutInterval: timeout))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(Body(
            model: model,
            messages: [.init(role: "system", content: system), .init(role: "user", content: user)],
            temperature: temperature, max_tokens: maxTokens
        ))
        return request
    }

    static func parseReply(_ data: Data) throws -> String {
        struct Reply: Decodable {
            struct Choice: Decodable { struct Message: Decodable { let content: String? }; let message: Message }
            let choices: [Choice]
        }
        if let message = serverError(in: data) { throw Failure.refused(message) }
        guard let reply = try? JSONDecoder().decode(Reply.self, from: data), let text = reply.choices.first?.message.content else {
            throw Failure.badReply(String(decoding: data.prefix(160), as: UTF8.self))
        }
        return text
    }

    static func parseModels(_ data: Data) throws -> [String] { try ModelCatalog.parse(data).map(\.id) }

    /// Both servers report trouble as `{"error": {"message": ...}}`.
    static func serverError(in data: Data) -> String? {
        struct Reply: Decodable { struct Error: Decodable { let message: String }; let error: Error }
        return (try? JSONDecoder().decode(Reply.self, from: data))?.error.message
    }

    private func authorised(_ request: URLRequest) -> URLRequest {
        var request = request
        if let apiKey { request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization") }
        if provider == .openRouter {
            // OpenRouter asks callers to say who they are; it shows up in their usage page.
            request.setValue("https://github.com/sworgkh/ledgelings", forHTTPHeaderField: "HTTP-Referer")
            request.setValue("Ledgelings", forHTTPHeaderField: "X-Title")
        }
        return request
    }

    // MARK: Network

    func listModels() async throws -> [String] { try await catalog().models.map(\.id) }

    /// Everything the provider offers, with names and prices where it gives them.
    func catalog() async throws -> ModelCatalog {
        ModelCatalog(models: try ModelCatalog.parse(try await fetch(authorised(URLRequest(url: modelsURL, timeoutInterval: 15)))))
    }

    func checkModel() async throws {
        let models = try await listModels()
        guard models.contains(model) else { throw Failure.modelMissing(model, available: models) }
    }

    /// OpenRouter only: the key's label and how much it has spent, or a refusal if the key is bad.
    func describeKey() async throws -> String {
        struct Reply: Decodable {
            struct Key: Decodable { let label: String?; let usage: Double?; let limit: Double? }
            let data: Key
        }
        let data = try await fetch(authorised(URLRequest(url: baseURL.appendingPathComponent("auth/key"), timeoutInterval: 10)))
        if let message = Self.serverError(in: data) { throw Failure.refused(message) }
        guard let key = try? JSONDecoder().decode(Reply.self, from: data).data else {
            throw Failure.badReply(String(decoding: data.prefix(120), as: UTF8.self))
        }
        var parts = [key.label ?? "key"]
        if let usage = key.usage { parts.append(String(format: "spent $%.2f", usage)) }
        if let limit = key.limit { parts.append(String(format: "of $%.2f", limit)) }
        return parts.joined(separator: " ")
    }

    /// One completion. `system` is who the speaker is; `user` is the moment.
    func reply(system: String, user: String, maxTokens: Int = 80, temperature: Double = 0.9) async throws -> String {
        try Self.parseReply(try await fetch(try request(system: system, user: user, maxTokens: maxTokens, temperature: temperature)))
    }

    private func fetch(_ request: URLRequest) async throws -> Data {
        do { return try await URLSession.shared.data(for: request).0 } catch { throw Failure.serverDown(error.localizedDescription) }
    }
}

File: win/Ledgelings/ChatClient.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Ledgelings.Core;

namespace Ledgelings;

/// <summary>
/// One language model behind an OpenAI-style chat endpoint. LM Studio on this
/// PC and OpenRouter on the internet both speak that dialect, so one client
/// serves both; only the address, the key and the model differ.
///
/// One trap this guards against: ask LM Studio for a model that is not
/// installed and it silently answers with whatever model happens to be loaded.
/// So <see cref="CheckModel"/> compares the id against the server's list before any
/// line is asked for.
/// </summary>
public sealed partial class ChatClient
{
    public enum Provider { LmStudio, OpenRouter }

    public static string Title(Provider provider) => provider == Provider.LmStudio ? "LM Studio" : "OpenRouter";

    public sealed class Failure : Exception
    {
        public Failure(string message) : base(message) { }
        public static Failure ServerDown(string why) => new("the server is not answering: " + why);
        public static Failure ModelMissing(string model, IReadOnlyList<string> available) =>
            new($"model {model} is not available" + (available.Count == 0 ? "" : " (have: " + string.Join(", ", available.Take(8)) + ")"));
        public static Failure Refused(string message) => new("the server refused: " + message);
        public static Failure BadReply(string what) => new("unexpected reply: " + what);
    }

    public const string OpenRouterUrl = "https://openrouter.ai/api/v1";

    private static readonly HttpClient http = new() { Timeout = TimeSpan.FromSeconds(60) };

    public Provider Kind { get; }
    /// <summary>The <c>/v1</c> root: <c>http://localhost:1234/v1</c> or <c>https://openrouter.ai/api/v1</c>.</summary>
    public string BaseUrl { get; }
    public string? ApiKey { get; }
    public string Model { get; }
    public string ProviderTitle => Title(Kind);

    private ChatClient(Provider kind, string baseUrl, string? apiKey, string model)
    {
        Kind = kind; BaseUrl = baseUrl.TrimEnd('/'); ApiKey = apiKey; Model = model;
    }

    public static ChatClient LmStudio(Uri server, string model) => new(Provider.LmStudio, server.ToString().TrimEnd('/') + "/v1", null, model);

    public static ChatClient OpenRouter(string key, string model) => new(Provider.OpenRouter, OpenRouterUrl, key, model);

    /// <summary>Enough to read OpenRouter's public model list; it needs no key.</summary>
    public static readonly ChatClient OpenRouterPublic = new(Provider.OpenRouter, OpenRouterUrl, null, "");

    public string ModelsUrl => BaseUrl + "/models";

    // MARK: Requests and replies, no network

    public HttpRequestMessage Request(string system, string user, int maxTokens = 80, double temperature = 0.9)
    {
        var body = new JsonObject
        {
            ["model"] = Model,
            ["messages"] = new JsonArray(
                new JsonObject { ["role"] = "system", ["content"] = system },
                new JsonObject { ["role"] = "user", ["content"] = user }),
            ["temperature"] = temperature,
            ["max_tokens"] = maxTokens,
        };
        // OpenRouter puts the price of the call in the reply when asked.
        if (Kind == Provider.OpenRouter) body["usage"] = new JsonObject { ["include"] = true };
        var request = Authorised(new HttpRequestMessage(HttpMethod.Post, BaseUrl + "/chat/completions"));
        request.Content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json");
        return request;
    }

    /// <summary>The text of a completion and, when the server reported it, what it used.</summary>
    public sealed record Answer(string Text, Spend.Usage? Usage);

    public static Answer ParseReply(string json)
    {
        if (ServerError(json) is string message) throw Failure.Refused(message);
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (!root.TryGetProperty("choices", out var choices) || choices.ValueKind != JsonValueKind.Array || choices.GetArrayLength() == 0)
                throw Failure.BadReply(Head(json, 160));
            var content = choices[0].GetProperty("message").TryGetProperty("content", out var c) && c.ValueKind == JsonValueKind.String ? c.GetString() : null;
            if (content is null) throw Failure.BadReply(Head(json, 160));
            Spend.Usage? usage = null;
            if (root.TryGetProperty("usage", out var u) && u.ValueKind == JsonValueKind.Object)
                usage = new Spend.Usage(
                    u.TryGetProperty("prompt_tokens", out var pt) && pt.TryGetInt32(out var p) ? p : 0,
                    u.TryGetProperty("completion_tokens", out var ct) && ct.TryGetInt32(out var q) ? q : 0,
                    u.TryGetProperty("cost", out var cost) && cost.ValueKind == JsonValueKind.Number ? cost.GetDouble() : null);
            return new Answer(content, usage);
        }
        catch (Exception e) when (e is JsonException or KeyNotFoundException or InvalidOperationException)
        {
            throw Failure.BadReply(Head(json, 160));
        }
    }

    public static List<string> ParseModels(string json) => ModelCatalog.Parse(json).Select(m => m.Id).ToList();

    /// <summary>Both servers report trouble as <c>{"error": {"message": ...}}</c>.</summary>
    public static string? ServerError(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.ValueKind == JsonValueKind.Object && doc.RootElement.TryGetProperty("error", out var error)
                && error.ValueKind == JsonValueKind.Object && error.TryGetProperty("message", out var m) && m.ValueKind == JsonValueKind.String)
                return m.GetString();
        }
        catch (JsonException) { }
        return null;
    }

    internal static string Head(string text, int length) => text.Length <= length ? text : text[..length];

    private HttpRequestMessage Authorised(HttpRequestMessage request)
    {
        if (ApiKey is not null) request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", ApiKey);
        if (Kind == Provider.OpenRouter)
        {
            // OpenRouter asks callers to say who they are; it shows up in their usage page.
            request.Headers.TryAddWithoutValidation("HTTP-Referer", "https://github.com/sworgkh/ledgelings");
            request.Headers.TryAddWithoutValidation("X-Title", "Ledgelings");
        }
        return request;
    }
}
